### ISSUE TITLE
Webhooks and subscriptions should support `Epic Link` changes

### ISSUE THREAD

**Issue by levb** [mattermost/mattermost-plugin-jira#339]
We currently do not support webhook events coming from Jira Server for updating an Epic Link for an issue.

When an epic link is assigned in an issue, Jira Server sends a webhook with event `issuelink_created`. We currently do not support this event, so the event is ignored and not processed. When an Epic Link is removed, it sends an `issuelink_deleted` event. When an Epic Link is changed from one Epic to another, both events are sent, with the `issuelink_deleted` event being the first one sent.

Jira Cloud sends these same events in the same scenarios, but also sends an event with type `jira:issue_updated` in all cases, which is why it currently works with Jira Cloud.

Here is a proposed solution:

- For Jira Server:
  - When we get an `issuelink_created` event,  make a post saying the Epic Link was changed to (value)
  - When we get an `issuelink_deleted` event, make a post saying the Epic Link (value) was removed from the issue

- For Jira Cloud:
  - (A) Handle the same way as Jira Server. And ignore the changes to Epic Link in `jira:issue_updated` events.
  OR
  - (B) Ignore `issuelink_created` and `issuelink_deleted` events, and continue to process `jira:issue_updated` events with Epic Link changes in them. This option makes for cleaner posts, because a change in Epic Link will result in one post instead of two, but involves some conditional logic based on Server vs Cloud.

Originally, https://mattermost.atlassian.net/browse/MM-18738

---
If you're interested please comment here and come [join our "Contributors" community channel](https://community.mattermost.com/core/channels/tickets) on our daily build server, where you can discuss questions with community members and the Mattermost core team. For technical advice or questions, please  [join our "Plugin: Jira" community channel](https://community.mattermost.com/core/channels/jira-plugin).

New contributors please see our [Developer's Guide](https://developers.mattermost.com/contribute/getting-started/) and our [Plugins Guide](https://developers.mattermost.com/extend/plugins/).

**Comment by levb** [comment]
👍 for option (B)

**Comment by codevbus** [comment]
I'd like to work on this.

**Comment by levb** [comment]
Thank you @codevbus for taking this on, feel free to join the conversation and ask for any help you might need at https://community-daily.mattermost.com/core/channels/jira-plugin.

**Comment by levb** [comment]
@codevbus Are you still interested in finishing this ticket?

**Comment by iamsayantan** [comment]
I would like to work on this.

**Comment by jfrerich** [comment]
Thanks @iamsayantan! If you have any questions, feel free to ask here or in the Mattermost channels listed in the issue description.  I'm removing the `Up For Grabs` label and this one is yours! :)

**Comment by iamsayantan** [comment]
Hi, I need a clarification. 

> When an epic link is assigned in an issue, Jira Server sends a webhook with event issuelink_created. We currently do not support this event, so the event is ignored and not processed. When an Epic Link is removed, it sends an issuelink_deleted event. When an Epic Link is changed from one Epic to another, both events are sent, with the issuelink_deleted event being the first one sent.
> 
> Jira Cloud sends these same events in the same scenarios, but also sends an event with type jira:issue_updated in all cases, which is why it currently works with Jira Cloud.

So I installed Jira server on my local machine and connected with Mattermost, and when I assigned an epic link to an issue I get both the `issuelink_created` and `jira:issue_updated` event in the webhook.

The `jira:issue_updated` event changelog is like this
```
"changelog": {
        "id": "10011",
        "items": [
            {
                "field": "Epic Link",
                "fieldtype": "custom",
                "from": null,
                "fromString": null,
                "to": "10000",
                "toString": "MAT-1"
            }
        ]
    }
```
which is being parsed by [`parseWebhookUpdatedField`](https://github.com/mattermost/mattermost-plugin-jira/blob/9fc652e107334f5122e90bf39edc35dea5bb0a1b/server/webhook_parser.go#L169) and generating post like this
![mattermost](https://user-images.githubusercontent.com/19422764/73137847-82a50380-4082-11ea-8fb4-f7a6222cb778.jpg)

But this is supposed to be the situation for Jira Cloud. I tried logging the Jira instance type and it prints `server`.
```
currentInstance, _ := p.currentInstanceStore.LoadCurrentJIRAInstance()
p.debugf("CurrentInstance", "INSTANCE", currentInstance.GetType())
```
Am I doing something wrong here?

**Comment by mickmister** [comment]
I may need to take a look at the behavior at both platforms again to verify that I have the two behaviors correct. What version of Jira server are you using @iamsayantan?

**Comment by iamsayantan** [comment]
I am using version `8.6.1`

**Comment by iamsayantan** [comment]
I am using version `8.6.1`

**Comment by mickmister** [comment]
@iamsayantan Sorry about the delay on this. I'll get back soon!

**Comment by iamsayantan** [comment]
Sure. No problem.

**Comment by levb** [comment]
@mickmister ping

**Comment by mickmister** [comment]
@iamsayantan I've verified the behavior you describe works with `Jira Cloud` and `Jira Server 8.x`.  I've determined that the behavior with `Jira Server` that I describe in the description is specifically for `Jira Server 7.x`, which this plugin currently supports.

Here is a sample `docker-compose.yml` that I use to run `Jira Server 7.13` locally.

```yaml
version: '3'
services:
  jira:
    image: atlassian/jira-software:7.13
    ports:
      - "8080:8080"
```

To be clear, the task here is to have the feature work properly for `Jira Server 7.x`, while also continuing to support the others. We can determine which version of `Jira Server` of being used by hitting the endpoint described [here](https://community.atlassian.com/t5/Answers-Developer-Questions/how-to-retrieve-installed-jira-version-via-REST-api/qaq-p/516360). We would only want to check if the installed version is `Jira Server`, and the event type is `issuelink_created` or `issuelink_deleted`.

@levb @aaronrothschild This situation is a bit of a corner case, but customers on `Jira Server 7.x` do not have working events regarding `Epic Link` as a result. 1/5 I think we should just move forward and not support this to avoid the complexity here, and tell customers to upgrade to `8.x` to have this functionality working properly for them.
